redfish-core/lib/led.hpp

```diff
@@ -239,11 +239,10 @@ inline void
 {
     BMCWEB_LOG_DEBUG << "Get LocationIndicatorActive";
 
-    nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];
     dbus::utility::getAssociationEndPoints(
         objPath + "/identifying",
```

Comment:
I think line 242 is actually a bug in led.hpp, where if getLocationIndicatorActive bails out for any reason (i.e. ec is set in the lambda function of that function or  in getLedAsset it's lambda function sees ec or ec1 errors), it will
leave a LocationIndicatorActive with a null value.  If one looks at the 1030 level of led.hpp, this behavior does
not appear to exist:  In that case jsonInput is passed straight thru to getLedAsset, which then does the following on if it succeeds: jsonInput["LocationIndicatorActive"] = assert . My change restored that one aspect of the 1030 led.hpp.

I can leave this behavior in led.hpp. I agree that it would be better to avoid changing code used by many others. However this then means the fabric_adapters.hpp needs to recognize this case and then do something like aResp->res.jsonValue.erase("LocationActiveIndicator). I actually looked at this option initially, but  aResp->res.jsonValue["LocationIndicatorActive"].is_null always was returning true even when this had a value of true/false. Do you know how I can correctly detect that this is null and then do the erase.

Reply:
As you explained, it looks like 1050 expects that the slot is already available for L236's getLocationIndicatorActive().
1030 creates a slot as necessary, on the other hand.


```
235 inline void
236     getLocationIndicatorActive(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
237                                const std::string& objPath,
238                                nlohmann::json& jsonInput)
239 {
240     BMCWEB_LOG_DEBUG << "Get LocationIndicatorActive";
241
242     nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];    .
243     dbus::utility::getAssociationEndPoints(
244         objPath + "/identifying",
...
258             getLedAsset(aResp, endpoint, jsonIn); <-- set the value to the above slot
...
261 }
262
```

Your analysis may be right in terms of the managing jsonslot.
However, I don't think the function `getLedAsset()` is defined to create/set the slot of 'LocationIndicatorActive'.
Even in 1030, the slot needs to already created before calling getLedAsset().

So, what can do may be.... (BUT THIS NEEDS VERIFICATION)

```
   getLocationIndicatorActive(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
                               const std::string& objPath,
                               nlohmann::json& jsonInput)
{
    BMCWEB_LOG_DEBUG << "Get LocationIndicatorActive";

242     nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];    .
==>
ADD-BEGIN:
    // Add new LocationIndicatorActive object to JSON array
    jsonInput.push_back({{"LocationIndicatorActive"}});
    nlohmann::json& jsonIn = jsonInput.back();
ADD-END:

243     dbus::utility::getAssociationEndPoints(
244         objPath + "/identifying",
...
258             getLedAsset(aResp, endpoint, jsonIn);  <== Use the above created slot
...
261 }
```

Reply:
The above change, seems to break bmcweb, where requests fail due to SSL_read errors.  indicating expected eof while reading.

Reply:
I think the syntax is incorrect as it attempts add a 2nd level array into jsonInput, and it may also attempt to add another element even if "LocationIndicatorActive" slot already exists.

It may be something like this:

```
242     nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];    .
==>

ADD>
   if( jsonInput.find("LocationIndicatorActive") == jsonInput.end())
   {
            jsonInput["LocationIndicatorActive"] = nullptr      <-- Add an element if not already existed.
    }
END-ADD>

    nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];
```

However, I don't think this is even needed.

--

The reason why it did not work in fabric, it does not find the associcationPoints.

```
(2023-03-21 13:39:18) [DEBUG "led.hpp":242] Get LocationIndicatorActive
(2023-03-21 13:39:18) [DEBUG "led.hpp":266] TEST: getAssociationEndPoints  NO /identifying
```

This is my test code to verify it

```
inline void
    getLocationIndicatorActive(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
                               const std::string& objPath,
                               nlohmann::json& jsonInput)
{
    BMCWEB_LOG_DEBUG << "Get LocationIndicatorActive";

    nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];
    dbus::utility::getAssociationEndPoints(
        objPath + "/identifying",
        [aResp, &jsonIn](const boost::system::error_code& ec,
                         const dbus::utility::MapperEndPoints& endpoints) {
        if (ec)
        {
            if (ec.value() != EBADR)
            {
                messages::internalError(aResp->res);
            }
            BMCWEB_LOG_DEBUG  << "TEST: getAssociationEndPoints  NO /identifying";  <== NO endpoints were found
            return;
        }

        BMCWEB_LOG_DEBUG << "TEST: getAssociationEndPoints  SETTING endPoints, size=" << endpoints.size();

        for (const auto& endpoint : endpoints)
        {
            getLedAsset(aResp, endpoint, jsonIn);
        }
        });
}
```

I think this may be because of 'objPath' which may not be correct.

Reply:
Trying it on rainer system (which has fabric adapter), it works without changing led.hpp at all.

```
(2023-03-21 15:14:43) [DEBUG "http_response.hpp":208] 0x1c06e38 setting completion handler
(2023-03-21 15:14:43) [DEBUG "led.hpp":242] Get LocationIndicatorActive /xyz/openbmc_project/inventory/system/chassis/motherboard/disk_backplane0
(2023-03-21 15:14:43) [DEBUG "led.hpp":272] TEST: getAssociationEndPoints  SETTING endPoints, size=1
```

I would suggest to try it again.

```
% curl -k -X GET https://${bmc}:18080/redfish/v1/Systems/system/FabricAdapters/chassis-motherboard-disk_backplane0
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/chassis-motherboard-disk_backplane0",
  "@odata.type": "#FabricAdapter.v1_4_0.FabricAdapter",
  "Id": "chassis-motherboard-disk_backplane0",
  "Location": {
    "PartLocation": {
      "ServiceLabel": "U78DB.ND0.WZS000G-P1"
    }
  },
  "LocationIndicatorActive": false,
  "Model": "6B89",
  "Name": "Fabric Adapter",
  "PartNumber": "02WG682",
  "Ports": {
    "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/chassis-motherboard-disk_backplane0/Ports"
  },
  "SerialNumber": "YA31UF09P00Z",
  "SparePartNumber": "02WG681",
  "Status": {
    "Health": "OK",
    "State": "Enabled"
  }
}%
```

Reply:
Those are not Fabric Adapters.
In 1030, PCIeDevice is not showing LocationIndicatorActive thru FabricAdapters.  
To get the LocationIndcatorActive for PCIeDevice, it needs to be done thru "PCIeSlots" , not even /redfish/v1/Systems/system/PCIeDevices/chassis_motherboard_pcieslot11_pcie_card11.


It can be checked using 1030 to see how it worked.
The remaining question (for those PCIeDevices thru FabricAdapters) may be --

- (a) should we not show "LocationIndcatorActive: null" at all like how it is done in 1030, or 
- (b) leave it as "LocationIndicatorActive:null".  I think it is still valid with null.

**Perhaps, for the consistency, we may need to do (a) so that LocationIndicatorActive itself is not showing if it does not support.
For it, the code of 'getLocationIndicatorActive()' may need to be changed so that the value can set only if it is returned.**

You can also try it out on rainer system which has the true FabricAdapter not as as a PCIeDevice, e.g. 
`/redfish/v1/Systems/system/FabricAdapters/chassis-motherboard-disk_backplane0`

Reply:
Note that this is happening because there is no association found as I showed  above.

```
"LocationIndicatorActive": null, <========== Here is the issue that I believe led.hpp is causing
```
Here, the question would be more like - what to do if there is no association as I showed.

```
inline void
    getLocationIndicatorActive(const std::shared_ptr<bmcweb::AsyncResp>& aResp,
                               const std::string& objPath,
                               nlohmann::json& jsonInput)
{
    BMCWEB_LOG_DEBUG << "Get LocationIndicatorActive";

    nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];
    dbus::utility::getAssociationEndPoints(
        objPath + "/identifying",
        [aResp, &jsonIn](const boost::system::error_code& ec,
                         const dbus::utility::MapperEndPoints& endpoints) {
        if (ec)
        {
            if (ec.value() != EBADR)
            {
                messages::internalError(aResp->res);
            }
            **BMCWEB_LOG_DEBUG  << "TEST: getAssociationEndPoints  NO /identifying";  <== NO endpoints were found**
            ret
````

Reply:
To get the consistent behavior with 1030, this can be fixed by modifying led.hpp.
I also think this should be applied as 1050 bugfix.

```
diff --git i/redfish-core/lib/led.hpp w/redfish-core/lib/led.hpp
index bf3924f4..c3cf3a2f 100644
--- i/redfish-core/lib/led.hpp
+++ w/redfish-core/lib/led.hpp
@@ -239,7 +239,6 @@ inline void
 {
     BMCWEB_LOG_DEBUG << "Get LocationIndicatorActive";

-    nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];
     dbus::utility::getAssociationEndPoints(
         objPath + "/identifying",
         [aResp, &jsonIn](const boost::system::error_code& ec,
@@ -253,6 +252,7 @@ inline void
             return;
         }

+        nlohmann::json& jsonIn = jsonInput["LocationIndicatorActive"];
         for (const auto& endpoint : endpoints)
         {
             getLedAsset(aResp, endpoint, jsonIn);

```

Reply:
If the PCIe slots are not true fabric adapters, then why do they show up under this request:

curl -k -H "X-Auth-Token: $token" -X  GET https://${bmc}/redfish/v1/Systems/system/FabricAdapters/
{
  "@odata.id": "/redfish/v1/Systems/system/FabricAdapters",
  "@odata.type": "#FabricAdapterCollection.FabricAdapterCollection",
  "Members": [
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/chassis-motherboard-disk_backplane0"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/chassis-motherboard-disk_backplane1"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot0-pcie_card0
"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot3-pcie_card3
"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot4-pcie_card4
"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot8-pcie_card8
"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot10-pcie_card10"
    },
    {
      "@odata.id": "/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot11-pcie_card11"
    }
  ],
  "Members@odata.count": 8,
  "Name": "Fabric Adapter Collection"
}

Does that imply the collection listed above is inaccurate?

Reply:
If you query it (eg. `/redfish/v1/Systems/system/FabricAdapters/motherboard-pcieslot10-pcie_card10`), it will say it is PCIeDevice.  I think this is a fabric adapter but it is on PCIeSlots.

Anyway, try my newly suggested code which should fix the logic and make it as same as 1030's expectation.

You also always compare it with 1030 firmware.

Reply:
Thanks, I verified that change to led.hpp also works for the pcie_cards of fabric adapters.

Reply:
I pushed the update led.hpp